pkg/reduce/readloop/readloop.go

```diff
@@ -183,13 +179,14 @@ func (rl *ReadLoop) Process(ctx context.Context, messages []*isb.ReadMessage) {
 
 		// close any windows that need to be closed.
 		wm := processor.Watermark(m.Watermark)
-		closedWindows := rl.aw.RemoveWindow(time.Time(wm))
+		closedWindows := rl.windower.RemoveWindows(time.Time(wm))
 		rl.log.Debugw("closing windows", zap.Int("length", len(closedWindows)), zap.Time("watermark", time.Time(wm)))
 
 		for _, cw := range closedWindows {
-			partitions := cw.Partitions()
+			kw := cw.(*keyed.KeyedWindow)
```

Comment:
why do we need this type conversion, can we avoid it? It is a little complex relationship 😄

Reply:
list is not typed and so any operations on node requires a cast.